prov/usnic/src/usdf_poll.c

```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright (c) 2016, Cisco Systems, Inc. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <fi.h>
+#include <fi_enosys.h>
+#include <fi_util.h>
+
+#include "usdf.h"
+#include "usdf_poll.h"
+
+static int usdf_poll_poll(struct fid_poll *pollset, void **context, int count)
+{
+	struct usdf_poll *_poll;
+	struct dlist_entry *item;
+	struct fid_list_entry *entry;
+	int ret = 0;
+	int copied = 0;
+	int pending;
+
+	USDF_TRACE_SYS(DOMAIN, "\n");
+
+	if (!pollset || !context) {
+		USDF_WARN_SYS(DOMAIN, "pollset and context can't be NULL.\n");
+		return -FI_EINVAL;
+	}
+
+	_poll = poll_ftou(pollset);
+
+	fastlock_acquire(&_poll->lock);
+
+	dlist_foreach(&_poll->list, item) {
+		entry = container_of(item, struct fid_list_entry, entry);
+		switch (entry->fid->fclass) {
+		case FI_CLASS_CQ:
+			pending = usdf_check_progress(entry->fid);
```

Comment:
Related note here for the future, we may want to shift our initial iteration point in the list on each call so that if the user puts, say, 4 CQs in the poll set and then calls `fi_poll(pollset, contexts, 2)` over and over again, they don't only get notifications about the first two CQs in the poll set.  I know that's not totally trivial with the existing `dlist_` APIs.